Training5/task1/src/App.js

```diff
@@ -0,0 +1,304 @@
+  [(state) => state.todos.data, (state) => state.filter],
+  (todos, filter) => {
+    const { name, completed } = filter ?? {};
+    const { ids, entity } = todos ?? {};
+
+    return {
+      ids: ids?.filter((id) => {
+        const todoData = entity?.[id];
+        const conditionCompleted =
+          completed === undefined || todoData?.completed === completed;
+        const conditionName = todoData?.name
+          .toLowerCase()
+          .includes(name?.toLowerCase());
+        return conditionCompleted && conditionName;
+      }),
+      entity,
+    };
+  }
+);
+
+function debounce(func, timeout = 300) {
+  let timer;
+  return (...args) => {
+    clearTimeout(timer);
+    timer = setTimeout(() => {
+      func.apply(this, args);
+    }, timeout);
+  };
+}
+
+function App() {
+  const dispatch = useDispatch();
+
+  const filter = useSelector((state) => state.filter);
+  const statusActionDefault = useSelector(
+    (state) => state.todos.statusActionDefault
+  );
+  const statusAction = useSelector((state) => state.todos.statusAction);
+
+  const [dataTodo, setDataTodo] = React.useState({
+    completed: false,
+    name: "",
+  });
+  const [editId, setEditId] = React.useState("");
+
+  const todos = useSelector((state) =>
+    selectorTodoByNameCompleteStatus(state, filter)
+  );
+  const isCreate = !editId;
+
+  const todoEditData = todos?.entity?.[editId];
+
+  React.useEffect(() => {
+    dispatch(fetchTodo());
+  }, []);
+
+  React.useEffect(() => {
+    switch (statusActionDefault.status) {
+      case STATUS_ACTION.fail:
+        toast(statusActionDefault.error);
+        break;
+      case STATUS_ACTION.idle:
+        setDataTodo({
+          completed: false,
+          name: "",
+        });
+        break;
+      default:
+        break;
+    }
+  }, [statusActionDefault]);
+
+  React.useEffect(() => {
+    if (editId && statusAction[editId]?.status === STATUS_ACTION.success) {
+      dispatch(setStatusAction({ id: editId, status: STATUS_ACTION.idle }));
+      setEditId("");
+      setDataTodo({ completed: false, name: "" });
+    }
+  }, [statusAction, editId]);
+
+  const removeTodoClick = (id) => {
+    let isConfirm = window.confirm("Are you sure delete item?");
+    isConfirm && dispatch(removeTodo({ id }));
+  };
+
+  const onChangeCompleted = () => {
+    setDataTodo({
+      ...dataTodo,
+      completed: !dataTodo.completed,
+    });
+  };
+
+  const onChangeName = (e) => {
+    setDataTodo({
+      ...dataTodo,
+      name: e.target.value,
+    });
+  };
+
+  const onClickContinue = () => {
+    if (!dataTodo.name) {
+      toast.error("Name is require");
+      return;
+    }
+    if (isCreate) {
+      dispatch(createTodo(dataTodo));
+    } else {
+      dispatch(editTodo(dataTodo));
+    }
+  };
+
+  const onClickCancel = () => {
+    setEditId("");
+    setDataTodo({ completed: false, name: "" });
+  };
+
+  const debounceChangeName = debounce((name) => {
```

Comment:
debounced function should not be created every render in this component, it can make your debounce function working incorrect, because debounce function should create one time, and use it many times